dhis2/dhis2-core
# feat: Add support for using JWT tokens

### Summary

This PR adds support for using JWT OpenID Connect bearer tokens for authenticating against the DHIS2 API endpoints.
In order to use the DHIS2 mobile clients with an OpenID Connect logged-in user, the mobile/external clients needs login and fetch access/refresh and ID tokens directly against the provider and then use the tokens as bearer tokens in subsequent requests against the DHIS2 API. In the OAuth2 terminology having this kind of feature set on the server is called being a [resource server](https://www.oauth.com/oauth2-servers/the-resource-server/).

We are using the [BearerTokenAuthenticationFilter](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/oauth2/server/resource/web/BearerTokenAuthenticationFilter.html) from Spring security as the authentication filter. If the request has an `Authorization` header with a `Bearer [JWT_TOKEN]`, it will be parsed and validated  against the configured OIDC provider(s), if config parameter `oidc.jwt.token.authentication.enabled` is enabled and `oauth2.authorization.server.enabled` is disabled.

Since the ["deprecated"](https://github.com/spring-projects/spring-security/wiki/OAuth-2.0-Migration-Guide) existing OAuth2 authorization server features can't co-exist with this new JWT authentication features, only one can be enabled at the same time. So in other words as of right now you have to chose from either having existing OAuth2 authorization support or new JWT token support. Since in most cases if you are using JWT tokens, it means you are using an external provider, having "local" authorization server support at the same time usually doesn't make sense.

The JWT token support build upon the existing [OIDC login/SSO support](https://github.com/dhis2/dhis2-docs/blob/master/src/commonmark/en/content/sysadmin/installation.md#openid-connect-oidc-configuration). In order to use the JWT tokens you need to configure an OIDC provider that will issue your JWT tokens according to this guide: [https://github.com/dhis2/dhis2-docs/blob/master/src/commonmark/en/content/sysadmin/installation.md#openid-connect-oidc-configuration](https://github.com/dhis2/dhis2-docs/blob/master/src/commonmark/en/content/sysadmin/installation.md#openid-connect-oidc-configuration)

### Automatic testing
JWT token support has unit tests in this test class: [JwtBearerTokenTest.java](https://github.com/dhis2/dhis2-core/blob/c781f610c881039194bf6e94238057261fd130a0/dhis-2/dhis-web/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/security/JwtBearerTokenTest.java)
In order to test a full "happy path" with JWT tokens we need to do considerable prework to generate a valid token that can be used without a lot of mocking. Most of the test code used to generate the tokens are taken from the new experimental [Spring-Authorization-Server](https://github.com/spring-projects-experimental/spring-authorization-server) project. This project is considered to be the heir to the existing deprecated authorization server, hence some included classes in this PR can possibly be refactored to use their future jars, when the project is ready for production. The main dependency that does the actuall JWT encode/decode work was already included as a dependency: ["Nimbus Jose JWT"](https://connect2id.com/products/nimbus-jose-jwt), (and is the same in Spring-Authorization-Server) 

### Manual testing
To do a full test on a live server you need to do the following:
* Configure an OIDC provider see: [openid-connect-oidc-configuration](https://github.com/dhis2/dhis2-docs/blob/master/src/commonmark/en/content/sysadmin/installation.md#openid-connect-oidc-configuration)
* Enable `oidc.jwt.token.authentication.enabled` and disable `oauth2.authorization.server.enabled` in the dhis.conf
* Run the "JWT/OIDC auth test" Android client in an emulator. Note: **As soon as the DHIS2 mobile client is updated with JWT support, this will be used instead.**
* Start the DHIS2 server
* Run the Android client and do the Authorization code flow against your configured provider and obtain a JWT ID token.
* Press the "Call DHIS2 API" button in the Android client and observe the returned request.

[[https://jira.dhis2.org/browse/DHIS2-10337](https://jira.dhis2.org/browse/DHIS2-10337)]
